src/main.c

```diff
@@ -2550,21 +2571,82 @@ static int open_listener(int domain, int type, int protocol, struct sockaddr *ad
     }
 
     /* bind */
-    if (bind(fd, addr, addrlen) != 0)
+    if (bind(fd, addr, addrlen) != 0) {
+        h2o_perror("bind failed");
         goto Error;
+    }
+
+    /* XXX we just assume for non TCP listeners we are not using the BPF
+     * filter
+     */
+    if (protocol != IPPROTO_TCP) {
+        int ret;
+        struct sockaddr_storage ss;
+        socklen_t len = sizeof(ss);
+        ret = getsockname(fd, (struct sockaddr *)&ss, &len);
+        if (ret != 0) {
+            fprintf(stderr, "getsockname failed on listener: %s\n", strerror(errno));
+        }
+    }
+
+    return fd;
+
+Error:
+    if (fd != -1)
+        close(fd);
+
+    return -1;
+}
+
+static int finish_open_listener(int fd, int apply_bpf_filter, int cpus)
+{
+    struct sockaddr_storage ss;
+    socklen_t sslen = sizeof(ss);
+    int ret = 0;
+
+    if (fd < 0)
+        goto Error;
+
+    ret = getsockname(fd, (struct sockaddr *)&ss, &sslen);
+    if (ret != 0) {
+        fprintf(stderr, "getsockname failed on listener: %s\n", strerror(errno));
+    }
+    if (ss.ss_family == AF_UNIX) {
+        fprintf(stderr, "dont finish AF_UNIX socket listener\n");
+        return 0;
+    } else if (ss.ss_family == AF_INET) {
+        fprintf(stderr, "will finish AF_INET socket\n");
+    } else if (ss.ss_family == AF_INET6) {
+        fprintf(stderr, "will finish AF_INET6 socket\n");
+    } else {
+        fprintf(stderr, "unknown family: %lu, skipping finish\n", (unsigned long)ss.ss_family);
+        return 0;
+    }
 
     /* TCP-specific actions */
-    if (protocol == IPPROTO_TCP) {
+    if (1) {
 #ifdef TCP_DEFER_ACCEPT
         { /* set TCP_DEFER_ACCEPT */
             int flag = 1;
-            if (setsockopt(fd, IPPROTO_TCP, TCP_DEFER_ACCEPT, &flag, sizeof(flag)) != 0)
+            if (setsockopt(fd, IPPROTO_TCP, TCP_DEFER_ACCEPT, &flag, sizeof(flag)) != 0) {
```

Comment:
Maebe we need to retain IPPROTO_TCP check for this setsockopt.

Reply:
with this logic being split into `finish_open_listener`, `protocol` is no longer available, i could infer it from the fd type being SOCK_STREAM, but currently only tcp sockets are finished. I can try to address together with a refactor eliminating the `need_finish` flag above